tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -0,0 +1,290 @@
+
+  Raises:
+    ValueError: If both `forwards` and `spots` are supplied or if neither is
+      supplied.
+    ValueError: If both `continuous_dividends` and `cost_of_carries` are
+      supplied or if neither is supplied
+  """
+  if (spots is None) == (forwards is None):
+    raise ValueError('Either spots or forwards must be supplied but not both.')
+  if (continuous_dividends is None) == (cost_of_carries is None):
+    raise ValueError('At most one of continuous_dividends and cost_of_carries '
+                     'may be supplied')
+
+  # For ease, not split out - cost of calculating eu call price for all small in
+  # comparison to american prices
+  eu_prices = vanilla_prices(
+    volatilities=volatilities, strikes=strikes, expiries=expiries, spots=spots,
+    forwards=forwards, discount_rates=discount_rates,
+    continuous_dividends=continuous_dividends, cost_of_carries=cost_of_carries,
+    dtype=dtype, name=name)
+
+  # Split out american
+  cost_of_carries = discount_rates - continuous_dividends \
+    if cost_of_carries is None else cost_of_carries
+
+  where_call = tf.constant(
+    True) if is_call_options is None else tf.reduce_any(is_call_options)
+  calculate_eu = tf.logical_and(where_call, cost_of_carries >= discount_rates)
+
+  if forwards is not None:
+    expiries = tf.convert_to_tensor(expiries, dtype=dtype, name="expiries")
+    spots = tf.convert_to_tensor(
+      forwards * tf.exp(-cost_of_carries * expiries), dtype=dtype, name="spots")
+  else:
+    expiries = tf.convert_to_tensor(expiries, dtype=dtype, name="expiries")
+    spots = tf.convert_to_tensor(spots, dtype=dtype, name="spots")
+
+  if is_call_options is not None:
+    is_call_options = tf.convert_to_tensor(is_call_options, dtype=tf.bool)
+  strikes = tf.convert_to_tensor(strikes, dtype=dtype, name="strikes")
+  volatilities = tf.convert_to_tensor(
+    volatilities, dtype=dtype, name="volatilities")
+  if continuous_dividends is not None:
+    cost_of_carries = tf.convert_to_tensor(
+      discount_rates - continuous_dividends, dtype=dtype,
+      name="cost_of_carries")
+  else:
+    cost_of_carries = tf.convert_to_tensor(
+    cost_of_carries, dtype=dtype, name="cost_of_carries")
+
+  discount_rates = tf.convert_to_tensor(
+    discount_rates, dtype=dtype, name="discount_rates")
+
+  am_prices, converged, failed = _option_price(
+    sigma=volatilities, x=strikes, t=expiries, s=spots, r=discount_rates,
+    b=cost_of_carries, is_call_options=is_call_options, dtype=dtype,
+    max_iterations=max_iterations, tolerance=tolerance, name=name)
+
+  return tf.where(calculate_eu, eu_prices, am_prices), converged, failed
+
+
+def _option_price(*,
+                  sigma,
+                  x,
+                  t,
+                  r,
+                  b,
+                  s=None,
+                  is_call_options=None,
+                  dtype=None,
+                  max_iterations=20,
+                  tolerance=1e-8,
+                  name=None):
```

Comment:
name here is not needed as this is the name of the adesi_whaley function

Reply:
its ok that we don't pass it through to vanilla prices?